### PR TITLE
Experiments on OpenJDK formulae build

### PR DESCRIPTION
**Pull request by unshare**
OpenJDK 8 and 11 are built with Xcode 14.3 as there exists legacy code incompatible with current default Clang behavior.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by SMillerDev** [comment]
Please split this into different pull requests since it'll be impossible to keep track of all three dependency trees and issues in one pull request.

In fact, https://github.com/Homebrew/homebrew-core/pull/160093 already exists for 21.0.2

**Comment by unshare** [comment]
@SMillerDev I'm investigating build failures with CI. It's a draft PR, please, disregard it until I finish checking my hypotheses.

File: Formula/o/openjdk@8.rb [hunk 1]
```diff
@@ -32,6 +32,10 @@ class OpenjdkAT8 < Formula
   uses_from_macos "unzip"
   uses_from_macos "zip"
 
+  on_ventura :or_newer do
+    disable! date: "2024-01-25", because: "is incompatible with Xcode 15+"
+  end
```
Comment (cho-m) [inline Formula/o/openjdk@8.rb:37]: What is exact issue? Do you have OpenJDK bug id?

You also mentioned OpenJDK 11 but builds/runs with upstream backport https://github.com/openjdk/jdk11u-dev/commit/dc028f28d1ec5a4efd89af1b5f83fa4dc349defc for https://bugs.openjdk.org/browse/JDK-8323008 (marked as affecting only 11, 17, 21, 22, 23).

---

Anyway, using `disable!` would be disruptive for users as it blocks installing the older bottle. Currently, Sonoma users are running off of the Ventura-built bottle, but this line would prevent installation.

Comment (unshare) [reply Formula/o/openjdk@8.rb:37]: That was just an experiment. Disregard it: this is still a draft PR.

**Comment by unshare** [comment]
@cho-m @SMillerDev
Feel free to object and criticize the following text. After all, I didn't restrain myself, so there exist several value judgements in the following review.
Creating this PR was necessary because the CI has so much more variability than I can realistically achieve on my single macOS device. Sorry for the compute resources expended.

**Turns out, OpenJDK 8 and 11 are incompatible with Xcode 15.**
* Both contain legacy C++ code that is not valid C++11. Previous compiler versions didn't consider it a big deal AND recent Autotools versions attempt to enable C++11 by default. Upstream keeps their Autotools version fixed, so configure scripts as is run fine. But if one patches stuff then re-runs `autogen.sh` or something similar (the Homebrew way), it's a bad idea from the beginning as it would likely break the build. Which is the case.
* OpenJDK 8 requires `JavaNativeFoundation` which is no longer required by modern Java and hence dropped in macOS 14. As macOS 13 builders use Xcode 15 and macOS SDK 14, building OpenJDK 8 on both macOS 13 and macOS 14 is kind of impossible (disabling GUI support is not an option). Only macOS 12 builds are successful because Xcode has dropped support for it while Homebrew hasn't. I tried to find a way use only macOS 12 for build and rely on backward compatibility, but it looks like Homebrew supports a lower bound on macOS release and has no provision for upper bound.

I've successfully built both OpenJDK 8 and 11 on my local macOS 13 machine using a different compiler and macOS 13 SDK.  
But running such a build of OpenJDK 11 using Homebrew is still a work in progress. I'll push it into this PR as soon as it's ready. (Title and description will be changed accordingly.)

As for OpenJDK 8, my proposal is to deprecate the OpenJDK 8 formula and resurrect OpenJDK 8 cask(s).  
It would be a separate endeavor.
The rationale/user story:
* Lots of enterprise customers run lots of legacy Java-based systems on Java 8 because of a) delayed Java 9, b) breaking changes in the following versions and c) generous long-term support.
* These technologically backward enterprises tend to outsource their Java development.
* Outsource companies tend to hire inexperienced developers to keep costs low. For the same reason, they promote a bring-your-own-device (BYOD) policy.
* Both said enterprises and said outsource companies rarely provide adequate DevOps infrastructure, so said inexperienced developers need to do everything on their localhosts. Setting up their localhost environment is often a burden of their more experienced piers and/or superiors.
* macBook Air/Pro happens to be perceived as a fashionable platform for Java developers: juniors waste money on macBooks due to a cargo cult while seniors run virtualized Linux on premium hardware (which makes sense).
* As a lead developer, I would like to streamline deployment of OpenJDK 8 to subordinate developers' machines. Homebrew seems to be a suitable tool for that.

If there still exist `openjdk@8` dependencies, it's okay, they can still use the outdated Java update that is available from Homebrew at the moment.
But if the user's story is somewhat resembling the bullets aforementioned, the cask would suit better because the real OpenJDK vendors tend to have their own CI unaffected by yearly macOS updates.

**Comment by SMillerDev** [comment]
> Turns out, OpenJDK 8 and 11 are incompatible with Xcode 15.

Can you report this to the OpenJDK project?

> I'll push it into this PR as soon as it's ready. (Title and description will be changed accordingly.)

Please don't do this, make a new pull request instead. Since, as I mentioned, it'll be impossible to keep track of ~all three~ both dependency trees and issues in one pull request.

> As for OpenJDK 8, my proposal is to deprecate the OpenJDK 8 formula and resurrect OpenJDK 8 cask(s).

Those casks already exist, there just won't be a dependency in Homebrew/core on them.

**Pull request closed without merging**